{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6e14ff64-edf3-49fb-96af-6deae5540300",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tensorflow version 2.7.0\n",
      "Num GPUs Available:  1\n",
      "Test set is  A0101\n"
     ]
    }
   ],
   "source": [
    "from __future__ import division, print_function, absolute_import\n",
    "import tensorflow as tf\n",
    "import numpy as np\n",
    "import math, os, time, sys, re, datetime\n",
    "from datetime import timedelta\n",
    "from sklearn.metrics import roc_auc_score, roc_curve, auc\n",
    "from scipy import stats\n",
    "\n",
    "print(\"Tensorflow version \" + tf.__version__)\n",
    "print(\"Num GPUs Available: \", len(tf.config.list_physical_devices('GPU')))\n",
    "\n",
    "#testset = 'A0202'\n",
    "#INPUTS: python DLMHC.py 0 A0202\n",
    "#testset = sys.argv[2] # e.g. takes A0202 as input allele name\n",
    "testset = \"A0101\"\n",
    "\n",
    "print(\"Test set is \", testset)\n",
    "runindx = 0 \n",
    "\n",
    "###all the possible sequence letters\n",
    "allSequences = 'ACEDGFIHKMLNQPSRTWVYZ'\n",
    "# Establish a mapping from letters to integers\n",
    "char2int = dict((c, i) for i, c in enumerate(allSequences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b9f78aee-9908-476e-86c3-5c2effea81e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getdata_onehot(datafile):   #build testing matrix\n",
    "    ### READ in test dataset\n",
    "    \"\"\" Reads the test data file and extracts allele subtype,\n",
    "            peptide length, and measurement type. Returns these information\n",
    "            along with the peptide sequence and target values.\n",
    "    \"\"\"\n",
    "    #train数据载入\n",
    "    import os\n",
    "    traindata = os.path.join(\"./../DATA\", \"train_data\", datafile )\n",
    "    print(\"traindata name: \", traindata)\n",
    "    df_train = pd.read_csv(traindata, header=0)\n",
    "    df_train = df_train[df_train.Peptide.str.contains('X') == False]\n",
    "    df_train = df_train[df_train.Peptide.str.contains('B') == False]\n",
    "    df_train = df_train[df_train.Peptide.str.contains('U') == False]\n",
    "    #eg.df_train = pd.read_csv('./DATA/train_data/A0202',sep=\"\\t\")\n",
    "    \n",
    "    #下采样\n",
    "    new_df_0 = df_train.loc[df_train['BindingCategory']== 0].sample(frac = 1)\n",
    "    #上采样\n",
    "    df_1_list = []\n",
    "    for i in range(4):\n",
    "        df_1_list.append(df_train.loc[df_train['BindingCategory']== 1])\n",
    "        new_df_1 = pd.concat(df_1_list)\n",
    "    new_df_train = pd.concat([new_df_0,new_df_1])\n",
    "    new_df_train = new_df_train.sample(frac = 1.0) #shuffle\n",
    "\n",
    "\n",
    "    #X_train--补齐11mer--one_hot_matrix\n",
    "    train_data=transformEL(new_df_train)\n",
    "    trainMatrix = np.empty((0, 11,len(allSequences)), int)      \n",
    "    for num in range(len(train_data.Peptide)):\n",
    "        if num%1000 == 0:\n",
    "            print(train_data.Peptide.iloc[num],num)\n",
    "        trainMatrix = np.append(trainMatrix, [Pept_OneHotMap(train_data.Peptide.iloc[num])], axis=0)\n",
    "    allele_name = train_data['HLA'][0]\n",
    "    assert (trainMatrix.shape[0] == train_data.shape[0])\n",
    "\n",
    "    #test数据载入\n",
    "    testdata = os.path.join(\"./../DATA\", \"test_data\", datafile )\n",
    "    df_test = pd.read_csv(testdata, header=0)\n",
    "    df_test = df_test[df_test.Peptide.str.contains('X') == False]\n",
    "    df_test = df_test[df_test.Peptide.str.contains('B') == False]\n",
    "    df_test = df_test[df_test.Peptide.str.contains('U') == False]\n",
    "    #eg.df_test = pd.read_csv('./DATA/test_data/A0202',sep=\"\\t\")\n",
    "\n",
    "    #X_test--补齐11mer--one_hot_matrix\n",
    "    test_data=transformEL(df_test)\n",
    "    testMatrix = np.empty((0, 11,len(allSequences)), int)      \n",
    "    for num in range(len(test_data.Peptide)):\n",
    "        if num%1000 == 0:\n",
    "            print(test_data.Peptide.iloc[num],num)\n",
    "        testMatrix = np.append(testMatrix, [Pept_OneHotMap(test_data.Peptide.iloc[num])], axis=0)\n",
    "    assert (testMatrix.shape[0] == test_data.shape[0])\n",
    "\n",
    "    Y_train = train_data.BindingCategory\n",
    "    Y_test = test_data.BindingCategory \n",
    "    #\n",
    "    Y_train = Y_train.reset_index(drop=True)\n",
    "    Y_test = Y_test.reset_index(drop=True)\n",
    "    #\n",
    "    trainlen = len(trainMatrix)\n",
    "    ss1 = list(range(trainlen))\n",
    "    rnd.shuffle(ss1)\n",
    "    #\n",
    "    valsize= int(1000) #Validation set size is 100 for validations dataset\n",
    "    X_val1 = trainMatrix[ss1[0:valsize]]\n",
    "    Y_val1 = Y_train.iloc[ss1[0:valsize]]\n",
    "    X_val2 = trainMatrix[ss1[valsize:(2*valsize)]]\n",
    "    Y_val2 = Y_train.iloc[ss1[valsize:(2*valsize)]]\n",
    "    X_val3 = trainMatrix[ss1[(2*valsize):(3*valsize)]]\n",
    "    Y_val3 = Y_train.iloc[ss1[(2*valsize):(3*valsize)]]    \n",
    "\n",
    "    trainMatrix = np.delete(trainMatrix,ss1[0:(3*valsize)], axis=0)\n",
    "    Y_train = Y_train.drop(Y_train.index[ss1[0:(3*valsize)]])\n",
    "    # combine training and test datasets\n",
    "    datasets={}\n",
    "    datasets['X_train'] = trainMatrix\n",
    "    datasets['Y_train'] = Y_train.values #traindata.BindingCategory.as_matrix()\n",
    "    datasets['X_test'] = testMatrix\n",
    "    datasets['Y_test'] = Y_test.values\n",
    "    datasets['X_val1'] = X_val1\n",
    "    datasets['Y_val1'] = Y_val1.values\n",
    "    datasets['X_val2'] = X_val2\n",
    "    datasets['Y_val2'] = Y_val2.values\n",
    "    datasets['X_val3'] = X_val3\n",
    "    datasets['Y_val3'] = Y_val3.values\n",
    "\n",
    "    return datasets\n",
    "\n",
    "def Pept_OneHotMap(peptideSeq):\n",
    "    \"\"\" maps amino acid into its numerical index\n",
    "    USAGE\n",
    "    Pept_OneHotMap('A')\n",
    "    array([[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])\n",
    "    \"\"\"\n",
    "    # integer encode input data\n",
    "    integer_encoded=[char2int[char] for char in peptideSeq]\n",
    "    # one hot encode\n",
    "    onehot_encoded = list()\n",
    "    for value in integer_encoded:\n",
    "    \tletter = [0 for _ in range(len(allSequences))]\n",
    "    \tletter[value] = 1\n",
    "    \tonehot_encoded.append(letter)\n",
    "    return np.asarray(onehot_encoded)\n",
    "\n",
    "def transformEL(dataset):\n",
    "    dataset = dataset.reset_index(drop=True)\n",
    "    peptide=dataset.Peptide\n",
    "    peptide2list=peptide.tolist()\n",
    "    for i in range(len(peptide)):\n",
    "        if len(peptide2list[i]) < 11:\n",
    "            n1 = len(peptide2list[i]) // 2\n",
    "            n2 = 11 - len(peptide2list[i])\n",
    "            peptide2list[i] = peptide2list[i][:n1] + 'Z'*n2 + peptide2list[i][n1:]     #将小于11个氨基酸的peptide在中间插空补齐\n",
    "        else:\n",
    "            peptide2list[i] = peptide2list[i][:11]\n",
    "\n",
    "    del dataset['Peptide']\n",
    "    peptides = pd.DataFrame(peptide2list,columns=['Peptide'])\n",
    "    dataset.insert(0,'Peptide',peptides.pop('Peptide'))\n",
    "    return dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "64548298-9183-41a6-bb0c-9e9fb609a0be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "traindata name:  ./../DATA/train_data/A0101\n",
      "RENATLTELRV 0\n",
      "FLDAZZFRTFY 1000\n",
      "LGSMZZTQKSD 2000\n",
      "PGDPZZRGGRE 3000\n",
      "EEQHZZEVSVN 4000\n",
      "QTDSKMLLGMY 5000\n",
      "DSSGLZLSCSE 6000\n",
      "LCRKZZELRQI 7000\n",
      "LLDTZZTEKYL 8000\n",
      "KHSMILLENQI 9000\n",
      "PLLGDILHQGT 10000\n",
      "ALGDGALFYFG 11000\n",
      "ASEAZZFVQKY 12000\n",
      "ACTFIZELQDS 13000\n",
      "RKTLQZKELKI 14000\n",
      "NGEEZZKSPPN 15000\n",
      "DYILSWYGNLS 16000\n",
      "SLFCZZVLSGL 17000\n",
      "PLAFRZDATSA 18000\n",
      "ESDSSLNSNVY 19000\n",
      "LQKQLZDEANR 20000\n",
      "LSQAZZDFLVC 21000\n",
      "EDIREEVDTFM 22000\n",
      "IPAFZZRGYDQ 23000\n",
      "RSITEZEVSEY 24000\n",
      "YTETZZVNHHY 25000\n",
      "KVENZZTRWDY 26000\n",
      "IGDTZZPIDTF 27000\n",
      "SATEMIWAVLA 28000\n",
      "FLDKSZGLQGY 29000\n",
      "AADTVZLATLY 30000\n",
      "FIIPNZVVKYS 31000\n",
      "SRNMIYITCHL 32000\n",
      "NTEKZZAVECY 33000\n",
      "WLSDIZPLVQP 34000\n",
      "KISKZZADIGA 35000\n",
      "GEARZZVRLKG 36000\n",
      "TVQTZZTLESL 37000\n",
      "PAAEPVPHLQT 38000\n",
      "LPGRDDNSYMY 39000\n",
      "VEDPQZGKNTV 40000\n",
      "ARSVPSIAAAT 41000\n",
      "QGRYENGCAYF 42000\n",
      "QSNMZZMAATL 43000\n",
      "QSDMGZREAAY 44000\n",
      "TSDTGZEGTVV 45000\n",
      "LTDDZZLHKQY 46000\n",
      "DLPDPZQWKSS 47000\n",
      "DLDFZZARQYY 48000\n",
      "FSQKZZSTLVY 49000\n",
      "ENTNZZIGSLF 50000\n",
      "LLDPKZYGMFR 51000\n",
      "KGYNSKFNLDM 52000\n",
      "ESKQKZFVLKT 53000\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'Peptide'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_26913/2797831260.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgetdata_onehot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdatafile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtestset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/tmp/ipykernel_26913/1144030059.py\u001b[0m in \u001b[0;36mgetdata_onehot\u001b[0;34m(datafile)\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0mtestdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"./../DATA\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"test_data\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdatafile\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m     \u001b[0mdf_test\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtestdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 41\u001b[0;31m     \u001b[0mdf_test\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_test\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf_test\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPeptide\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontains\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'X'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     42\u001b[0m     \u001b[0mdf_test\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_test\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf_test\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPeptide\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontains\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'B'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     43\u001b[0m     \u001b[0mdf_test\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_test\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf_test\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPeptide\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontains\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'U'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/tf2-gpu/lib/python3.8/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5581\u001b[0m         ):\n\u001b[1;32m   5582\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5583\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5584\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5585\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'Peptide'"
     ]
    }
   ],
   "source": [
    "data=getdata_onehot(datafile=testset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ca782c12-b39d-4ac6-95db-738bf251b398",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'data' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_26913/677450282.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'X_train'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'data' is not defined"
     ]
    }
   ],
   "source": [
    "print(data['X_train'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a60e157b-63b0-4cf2-a529-32a74b0042e9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
